_posts/2024-10-08-Internal-functioning-of-sorting-in-hybrid-query.md

```diff
@@ -0,0 +1,122 @@
+
+## Query execution workflow
+
+At a high level, query execution is divided into two main phases: ***query*** and ***fetch***. During the query phase, OpenSearch retrieves matching results from each shard as objects containing document IDs and relevance scores. These results then move to the fetch phase, where the full text for each document ID is retrieved from the shards. Finally, OpenSearch combines the results from all shards, generates the final search response, and sends this response to the user. Query processing is discussed in more detail in [this blog post](https://opensearch.org/blog/a-query-or-there-and-back-again/). 
+
+## High-level logic of sorting in a hybrid query
+
+In a typical hybrid query, results are retrieved based on both lexical and semantic relevance scores. The top-scoring documents from each shard are then combined to form the final result, prioritizing relevance over any other sorting criteria.
+
+To enable sorting in this process, subquery results are retrieved according to the specified sort criteria. These sorted subquery results are then merged at the coordinator level, ensuring that the final output follows the user’s requested sort order rather than relying solely on relevance scores.
+
+The following example illustrates the sorting workflow in a hybrid query. In this case, a user triggers a hybrid query with two subqueries: `match` and `term`, and requests that the search results be ordered by stock price in descending order.
+
+First, the hybrid query executes the `match` and `term` subqueries separately, sorting each result set by the specified criterion—in this case, stock price. Unlike traditional sorting, which disregards relevance scores, hybrid queries include both the relevance score and the sorting fields in the results of query phase. These values are then used in the normalization process. During normalization, relevance scores are used to calculate normalization scores, which act as tie-breakers when subquery results are combined. Finally, the subquery results are combined at the coordinator level to form the final result, ensuring that the output is sorted by stock price in descending order, as requested.
+
+The following diagram illustrates the flow of the query phase and creating the final shard result based on sorting criteria.
+
+![hybrid-query-with-sorting](/assets/media/blog-images/2024-10-08-internal-functioning-of-sorting-in-hybrid-search/Hybrid-query-with-sorting.png){:class="img-centered"}
+
+
+## Detailed overview of sorting in a hybrid query 
+
+Building on the high-level logic discussed earlier, let's now dive into the detailed process of sorting in a hybrid query. This can be broken down into two key steps:
+
+1. Sorting individual subquery results at the shard level.
+2. Merging the multiple subquery results based on the sorting criteria at the coordinator node.
```

Comment:
```suggestion
2. Merging the multiple subquery results, based on the sorting criteria, at the coordinator node level.
```

Reply:
This refers to the coordinator node itself, so leaving as is.